pkgs/development/beam-modules/mix-release.nix

```diff
@@ -0,0 +1,92 @@
+{ stdenv, lib, elixir, erlang, hex, rebar, rebar3, fetchMixDeps, makeWrapper, git }:
+
+{ pname
+, version
+, src
+, nativeBuildInputs ? [ ]
+, meta ? { }
+, enableDebugInfo ? false
+, mixEnv ? "prod"
+, compileFlags ? [ ]
+, mixDeps ? null
+, ...
+}@attrs:
+let
+  overridable = builtins.removeAttrs attrs [ "compileFlags" ];
+
+in
+stdenv.mkDerivation (overridable // {
+  dontStrip = true;
+
+  inherit pname version src;
+
+  nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir makeWrapper git ];
+
+  MIX_ENV = mixEnv;
+  MIX_DEBUG = if enableDebugInfo then 1 else 0;
+  HEX_OFFLINE = 1;
+  DEBUG = if enableDebugInfo then 1 else 0; # for Rebar3 compilation
+  # the api with `mix local.rebar rebar path` makes a copy of the binary
+  MIX_REBAR = "${rebar}/bin/rebar";
+  MIX_REBAR3 = "${rebar3}/bin/rebar3";
+
+  postUnpack = ''
+    export HEX_HOME="$TEMPDIR/hex"
+    export MIX_HOME="$TEMPDIR/mix"
+    # compilation of the dependencies will require
+    # that the dependency path is writable
+    # thus a copy to the TEMPDIR is inevitable here
+    export MIX_DEPS_PATH="$TEMPDIR/deps"
+
+    # Rebar
+    export REBAR_GLOBAL_CONFIG_DIR="$TEMPDIR/rebar3"
+    export REBAR_CACHE_DIR="$TEMPDIR/rebar3.cache"
+
+    ${lib.optionalString (mixDeps != null) ''
+      cp --no-preserve=mode -R "${mixDeps}" "$MIX_DEPS_PATH"
+    ''
+    }
+
+  '' + (attrs.postUnpack or "");
+
+  configurePhase = attrs.configurePhase or ''
+    runHook preConfigure
+
+    # this is needed for projects that have a specific compile step
+    # the dependency needs to be compiled in order for the task
+    # to be available
+    # Phoenix projects for example will need compile.phoenix
+    mix deps.compile --no-deps-check --skip-umbrella-children
+
+    runHook postConfigure
+  '';
+
+  buildPhase = attrs.buildPhase or ''
+    runHook preBuild
+
+    mix compile --no-deps-check ${lib.concatStringsSep " " compileFlags}
+
+    runHook postBuild
+  '';
+
+
+  installPhase = attrs.installPhase or ''
+    runHook preInstall
+
+    mix release --no-deps-check --path "$out"
+
+    runHook postInstall
+  '';
+
+  postFixup = ''
```

Comment:
Can we somehow keep `postFixup` and `preFixup` untouched, such that we can use it to do our own fixups?

Also there is a file called `start` somewhere in the generated release, which is a shell script with a variable set `ROOTDIR` to the "systems" erlang installation, which again results in the system erlang kept as a runtime dependency in the closure.

It seems to be safe to just substitute it with an empty string.

I do it like this in my builds:

https://github.com/NobbZ/ts_helper/blob/a4f3b5631eaa6689cb1a1e8e080fe36ee4b84585/flake.nix#L82-L85

I'm going back and forth in hardcoding `disallowedReferences` to disallow the erlang runtime used for building in this builder in general.

Reply:
There is one hard part about using fixup. We need to not forget to run the steps that are necessary in fixup ourselves.
As far as I understand it, it's just a matter of using patchShebangs.
The other steps being stripping and patchelf. We do not strip and I don't know what patchelf is doing. So I will just use patchShebangs. Let me know if something else should be run in fixupPhase.

As for the ROOTDIR thing, it seems correct, however since we are not sure about it, how about leaving it in the postFixupPhase for those who are willing to experiment. I've check the official doc and that ROOT_DIR var isn't mentioned, so I would personally not like to depend on it
https://hexdocs.pm/mix/master/Mix.Tasks.Release.html

Reply:
I just tested it with replacing the fixupPhase and it all seems to work.

Reply:
I have never used `disallowedReferences` before and therefore I am not exactly sure when it kicks in.

Reply:
Here is what I also tried.
- using `rg` I couldn't find any related reference
- using 
```
❯ nix path-info -S ./result
/nix/store/023463n9h5alv79zgqz78d195y5bvpxl-union-0.0.1   395793744
```
it looks like the resulting derivation is about the same size as the original one. Maybe I misunderstood something, but I thought the point of this patch was to reduce the size of the derivation. Maybe nix path-info does not give you the size including the references?
Or perhaps, when installing the resulting derivation, with this reference it will require the target machine to have erlang installed?
Let me add that as a TODO then. I'll make further testing once the PR is merged.

Reply:
I've added a TODO with your potential fix. Let me know if I misunderstood anything.